### DISCUSSION TITLE
Stream Mode Mod

### DISCUSSION THREAD

**Discussion by 59xa** [ppy/osu#13075] [Ideas]
**Inspired by osu!stream’s Stream Mode.**
Basically, player starts on the intermediate difficulty of the beatmap and whenever they fail/pass that difficulty, the difficulty will increase/decrease. The game will tell you if you’re going to stream up or stream down. The game ends if player fails on the easy difficulty and difficulty increases until it reaches the limit of beatmap’s difficulty and stays there until the end or player fails to maintain that difficulty.
**How it should work:**
- Beatmap difficulty select will be restricted if mod is equipped
- Mods are allowed
- Should be unranked and placed in the fun section
- HP bar should start in the middle instead of full

**Comment by peppy** [comment]
The problem with this is that osu!stream was only playable because every beatmap was meticulously crafted to support stream switching. Doing this with normal maps will likely not feel good, unless you restrict difficulty changes to breaks only.

**Comment by 59xa** [comment]
>Doing this with normal maps will likely not feel good, unless you restrict difficulty changes to breaks only.

@peppy
Yeah, I also see the problem here as well. If we make difficulty changes on break, it will also not work due to the fact that most beatmaps don’t have it. Maybe have the difficulty change based on BPM tick? Would need something more to make it work pretty much.

**Comment by cl8n** [comment]
that might solve rhythm problems when switching streams (sometimes), but coordinating object placement across all difficulties in a set is something that would require explicit effort from the mapper, and it definitely wouldn't work for 99% of maps made so far

that being said it'd be awesome to see someone try to make maps that support this, and it'd probably work out-of-the-box better with taiko and mania?

**Comment by iSimpFor** [comment]
i think we can make it transition without being too jarring by putting a buffer

how i think it should transition(in osu and maybe osu!ctb):
it should remove all combo sets(starts with a new combo, and all the circles after it without a new combo) in the 4 ticks buffer when transitioning from easy to normal diffs
all combo sets 2 ticks from normal to hard
1 tick from hard to insane
1/2 a tick from insane to expert
no buffer for expert to expert+

**Comment by iSimpFor** [comment]
i have even more ideas for this mod

i suggest there be 2 versions of this mod, one is called stream normal, the other one will be called stream mix


for both, no hp drain and when transitioning, if 2 objects are on the same tick in time, the object for the difficulty before the transition will be removed, this means all objects
to transition to another difficulty:

for osu!std, osu!ctb, and osu!mania, hp needs to be either full or empty to transition

for osu!taiko, the hp system will be adjusted to work like the other gamemodes to allow it to work with this mod


for stream normal, all the above will apply and also instead of restricting difficulty options, it can either be that you can choose which difficulty to start and you will play stream mode from that difficulty instead of restricting difficulties or you can choose the difficulty you will start in depending on the highest difficulty you have cleared, the highest difficulty you have gotten to on stream mode, or based on recommended difficulty given to you by the game


for stream mix, it mixes all gamemodes of a mapset together
to handle the transitions:
when transitioning at all, the gamemode that is being transitioned to will not accept inputs until the transition to the gamemode happens

when transitioning to osu!std, a transparent object will appear one measure or 4 ticks before the transition to that mode which indicates the first object to appear at that location

when transitioning to osu!taiko or osu!mania, the lanes will appear and objects will start coming to the input points one measure or 4 ticks before the transition happens

when transitioning to osu!ctb, the catcher will appear under where the next fruit will land one measure or 4 ticks before the transition happens
